### ISSUE TITLE
Bug in OneOrMore/Or

### ISSUE THREAD

**Issue by geuben** [pyparsing/pyparsing#34]
I am seeing some potentially odd behaviour, or it's just me misunderstanding something.

I've narrowed it down to this trivial example (that on it's own doesn't make a lot of sense but it's part of something bigger)

```
import pyparsing as pp
a = pp.OneOrMore(
    pp.Or([
        pp.White(exact=1),
    ]))

expr1 = pp.White(exact=1) + pp.ZeroOrMore(pp.Word(pp.printables))
expr2 = a + pp.ZeroOrMore(pp.Word(pp.printables))

print(expr1.parseString(" foobar"))
print(expr2.parseString(" foobar"))
```

`expr1` works and returns `[' ', 'foobar']` (I'm not actually interested in the output)
`expr2` fails with:

`ParseException: Expected {<SPC><TAB><CR><LF>} (at char 1), (line:1, col:2)`

Doesn't expression `a` evaluate to the same as `pp.White(exact=1)` in the case of " "?

The reason for expression `a` is that in my actual code the `Or` contains some other expressions for matching punctuation.

**Comment by ptmcg** [comment]
The introduction of Or is causing pyparsing to perform its default behavior of skipping whitespace, so the leading space gets skipped before trying to match White. You can fix the problem by adding `leaveWhitespace()`:

    a = pp.OneOrMore(
        pp.Or([
            pp.White(exact=1),
        ])).leaveWhitespace()

But the pyparsing philosophy is for grammars to *not* explicitly include whitespace elements, but to leave that to pyparsing's internal indexing logic. And I know you said the Or was something of a contrived element, and in fact, an Or containing just a single expression really is unnecessary; but please use the `'^'`, `'|'`, `'+'`, etc. operators instead of building up expressions by passing lists to Or, MatchFirst, And, etc. respectively.

**Comment by geuben** [comment]
Can you clarify this point:

> but please use the '^', '|', '+', etc. operators instead of building up expressions by passing lists to Or, MatchFirst, And, etc. respectively.

Are you saying I should be doing `foo = 'a' | 'b' | 'c'` instead of `foo = MatchFirst(['a', 'b', 'c'])`. If so, why?

**Comment by ptmcg** [comment]
This is ultimately a style choice, as either form will certainly *work*, and functionally both are equivalent. I think the use of the operators makes the code easier to write and easier for others to grasp quickly, as it makes the code more similar to the abstract Backus-Naur Form grammar syntax. Being able to "read" the grammar from the Python code is important in pyparsing code. I used this same reasoning when adding support for `expr("attr_name")` as an equivalent alternative to `expr.setResultsName("attr_name")` - I believe the reader's eye can acquire the concept more quickly in the shortened form, and the brain does not lose its place in the larger context of the grammar in the parse expression.

The explicit class construction is exactly what one would have to write in a language that does not support operator overloading (Java comes to mind) - I believe Python's support for operator overloading is an important part of its API design, and so I embraced it when developing pyparsing. It *feels* more Pythonic, if I can invoke that nebulous term. The *only* time I ever construct a MatchFirst or And using the class constructor is when the component expressions are being built in a generator expression.

I'm also influenced from my prior years in development, even back to working in COBOL. Early versions of the language required writing `y = m*x + b` as:

    MOVE X TO Y
    MULTIPLY Y BY M
    ADD B TO Y.

The language was mercifully enhanced to support:

    COMPUTE Y = M * X + B.

Both are equivalent, but the verbosity of the first obscures the basic intent.

And in the end, I am also a lazy developer, so the operator form is easier for me to use in my own work. But as I said at the start, both forms are functionally equivalent, so use whichever you are comfortable with.
